src/main/gov/nasa/jpf/vm/Types.java

```diff
@@ -1070,30 +1070,18 @@ public static int hiLong (long l) {
     return (int) (l >> 32);
   }
 
-  public static boolean instanceOf (String type, String ofType) {
-    int bType = getBuiltinTypeFromSignature(type);
-
-    if ((bType == T_ARRAY) && ofType.equals("Ljava/lang/Object;")) {
-      return true;
+  public static boolean instanceOf (ClassInfo ci, String ofTypeSignature) {
+    int bOfType = getBuiltinTypeFromSignature(ofTypeSignature);
+    if (ci.isArray() && bOfType == T_ARRAY) {
+      return instanceOf(ci.getComponentClassInfo(), getArrayElementType(ofTypeSignature));
     }
 
-    int bOfType = getBuiltinTypeFromSignature(ofType);
-
-    if (bType != bOfType) {
-      return false;
+    if (ci.isPrimitive()) {
+      int bType = getBuiltinTypeFromSignature(ci.getSignature());
+      return bType == bOfType;
     }
```

Comment:
This special case for `ci.isPrimitive()` shouldn't be necessary, but I was confused as to what `ClassInfo.isInstanceOf` expects as argument:
https://github.com/javapathfinder/jpf-core/blob/8ae607ae65e3f5e92a04963157ee5315202ecefc/src/main/gov/nasa/jpf/vm/ClassInfo.java#L1620-L1629

When the `ClassInfo` is not primitive (the `else` branch), this method seems to expect `cname` to be a class name or a type name (as exemplified in the usage `ci.isInstanceOf(getTypeName(ofType))` in the old code). But for a primitive type (the `if` branch), it expects `cname` to be a type signature (e.g., `I` rather than `int`). So either way I had to special case it.